### ISSUE TITLE
Assembly loaded a second time in default host context by RuntimeAssembly.GetExportedTypes

### ISSUE THREAD

**Issue by MansenC** [dotnet/runtime#82856]
### Description

This is an issue I have reported separately in [the MSBuild repository](https://github.com/dotnet/msbuild/issues/8514) since I am not really sure where it fits exactly.

I am experiencing an issue where the assembly `Microsoft.Build.Framework` is being loaded twice into the same host context. It's the exact same assembly, same path and thus same properties and types. The twice-loaded assembly only appears when I start my application using a custom native host, and not when I run it in the debugger using Visual Studio, which makes this quite odd.

### Reproduction Steps

Update: I have made a repository reproducing the issue. It is found [here](https://github.com/MansenC/TwiceLoadedAssemblyIssue). Note that this is implemented to work only on Windows since I use a Windows machine.

I am not entirely sure how to reproduce this issue. I expect it does arise from the following steps: Create a native host and a C# application. I use `hostfxr_initialize_for_dotnet_command_line_fn` to "start" the application and then call C# functions using `load_assembly_and_get_function_pointer_fn`. That C# code in turn calls MSBuild functionality using MSBuild provided by the NuGet package `Microsoft.Build.Runtime`, in my case I have a function that starts a build using `BuildManager.DefaultBuildManager` (This function is where the first assembly load implicitly occurrs), and a separate one where I try to load an MSBuild Project (This constructor call is where the second load occurs after several other things).

MSBuild fails to restore a project due it not being able to cast a task called `WarnForInvalidProject` to `Microsoft.Build.Framework.ITask`. This class does implement that interface though, which makes me in turn suspect that the twice-loaded assembly causes that issue. If not, I do not think that this is intended behaviour.

### Expected behavior

No assembly is being loaded twice into the primary host context

### Actual behavior

The assembly is being loaded twice

### Configuration

I am using .NET Core 6.0.201 with Visual Studio 22 and the following MSBuild (-related) NuGet packages:

    Microsoft.Build.Locator 1.4.1
    Microsoft.Build.NuGetSdkResolver 5.11.0-rc.10
    Microsoft.Build.Runtime 17.3.2
    NuGet.Build.Tasks 6.5.0

All this on Windows 11 x64.

### Other information

I have located the two calls from which the assembly is being loaded. This being once, implicitly, when I first use a class that in turn uses the namespace `Microsoft.Build.Framework`, thus forcing a load of said assembly. The stacktrace after the call to that function looks as follows:
```
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at Internal.Runtime.InteropServices.IsolatedComponentLoadContext.Load(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingLoad(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.Resolve(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
   ```
This is most likely correct as it is where the assembly is loaded when I use Visual Studio. The erroneous load is made, as noted above, when I instantiate the class `Microsoft.Build.Evaluation.Project`. This calls a bunch of stuff and in the end uses `Assembly.ExportedTypes`, where the assembly load is made. This already sounds odd to me, since an instance of that assembly is required in order to get the exported types, so an assembly load there makes little sense to me. Yet again, I don't know how that all is implemented in the CoreCLR. I have found the source code for that but I'm no expert and never have worked with that code, so I don't really know what could cause that. All I know is that somewhere in there, a call is made that loads the assembly again. Here's the stacktrace:
```
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Reflection.Assembly.get_ExportedTypes()
   at Microsoft.Build.BackEnd.SdkResolution.SdkResolverLoader.GetResolverTypes(Assembly assembly)
   at Microsoft.Build.BackEnd.SdkResolution.SdkResolverLoader.LoadResolvers(String resolverPath, LoggingContext loggingContext, ElementLocation location, List`1 resolvers)
   at Microsoft.Build.BackEnd.SdkResolution.SdkResolverLoader.LoadResolversFromManifest(SdkResolverManifest manifest, LoggingContext loggingContext, ElementLocation location)
   at Microsoft.Build.BackEnd.SdkResolution.SdkResolverService.GetResolvers(IList`1 resolversManifests, LoggingContext loggingContext, ElementLocation sdkReferenceLocation)
   at Microsoft.Build.BackEnd.SdkResolution.SdkResolverService.ResolveSdkUsingResolversWithPatternsFirst(Int32 submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, String solutionPath, String projectPath, Boolean interactive, Boolean isRunningInVisualStudio)
   at Microsoft.Build.BackEnd.SdkResolution.SdkResolverService.ResolveSdk(Int32 submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, String solutionPath, String projectPath, Boolean interactive, Boolean isRunningInVisualStudio)
   at Microsoft.Build.BackEnd.SdkResolution.CachingSdkResolverService.<>n__0(Int32 submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, String solutionPath, String projectPath, Boolean interactive, Boolean isRunningInVisualStudio)
   at Microsoft.Build.BackEnd.SdkResolution.CachingSdkResolverService.<>c__DisplayClass3_0.<ResolveSdk>b__2()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at Microsoft.Build.BackEnd.SdkResolution.CachingSdkResolverService.ResolveSdk(Int32 submissionId, SdkReference sdk, LoggingContext loggingContext, ElementLocation sdkReferenceLocation, String solutionPath, String projectPath, Boolean interactive, Boolean isRunningInVisualStudio)
   at Microsoft.Build.Evaluation.Evaluator`4.ExpandAndLoadImportsFromUnescapedImportExpressionConditioned(String directoryOfImportingFile, ProjectImportElement importElement, List`1& projects, SdkResult& sdkResult, Boolean throwOnFileNotExistsError)
   at Microsoft.Build.Evaluation.Evaluator`4.ExpandAndLoadImports(String directoryOfImportingFile, ProjectImportElement importElement, SdkResult& sdkResult)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.Evaluate()
   at Microsoft.Build.Evaluation.Project.ProjectImpl.Reevaluate(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project.ProjectImpl.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project.ProjectImpl.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project.ProjectImpl.ReevaluateIfNecessary(EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project.ProjectImpl.Initialize(IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project..ctor(ProjectRootElement xml, IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectCollection projectCollection, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext, IDirectoryCacheFactory directoryCacheFactory)
   at Microsoft.Build.Evaluation.Project..ctor(ProjectRootElement xml, IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectCollection projectCollection, ProjectLoadSettings loadSettings)
   at Microsoft.Build.Evaluation.Project..ctor(ProjectRootElement xml, IDictionary`2 globalProperties, String toolsVersion, ProjectCollection projectCollection, ProjectLoadSettings loadSettings)
   at Microsoft.Build.Evaluation.Project..ctor(ProjectRootElement xml, IDictionary`2 globalProperties, String toolsVersion, ProjectCollection projectCollection)
```
Both do call `AssemblyLoadContext::OnAssemblyLoad`, where I compare the assembly name and print the stacktrace using `Environment.StackTrace`. But that's where my ability to find useful information ends.

In the mean time I'm going to try and set up a working project where this issue arises, but I expect it not being that "minimal". Thank you in advance

**Comment by vitek-karas** [comment]
> This already sounds odd to me, since an instance of that assembly is required in order to get the exported types, so an assembly load there makes little sense to me.

This could be a situation where assembly A has an exported type (type forward) which redirects to assembly B. Enumerating the exported types on A will likely cause the runtime to resolve the reference to B, which will cause assembly load event for B. This is probably happening from a secondary ALC and it's the first time B was seen in that ALC. So this runs the usual assembly resolution logic. I know that MSBuild implements relatively complicate resolution logic for the secondary ALCs. It's possible that the difference in hosting causes this logic to behave differently.

For example it's possible that some environment will look slightly differently in the native hosting since normal app startup didn't happen. 

@rainersigwald might have an idea, or at least things to look for.

/cc @elinor-fung

**Comment by elinor-fung** [comment]
The assembly is being loaded into two different [`AssemblyLoadContext`](https://learn.microsoft.com/dotnet/core/dependency-loading/understanding-assemblyloadcontext) instances - the default context (from the explicit call to `AssemblyLoadContext.Default.LoadFromAssemblyPath`) and an isolated context. You can check what ALC an assembly is in using [`AssemblyLoadContext.GetLoadContext`](https://learn.microsoft.com/dotnet/api/system.runtime.loader.assemblyloadcontext.getloadcontext).

`load_asembly_and_get_function_pointer` loads the assembly in an isolated context that is set up to resolve dependencies using the `.deps.json` (dependencies manifest) next to the assembly. In the repro, `UseMSBuildFrameworkClass` is running in that isolated context so the reference to `BuildStartedEventArgs` results in loading `Microsoft.Build.Framework` - found in the assembly's `.deps.json` - in that isolated context. 

Using `hostfxr_run_app` / `get_function_pointer` will run the app / look for the function pointer in the default ALC, which is why you would not see a second load in an isolated context.

Are you only calling into the one assembly that is used to `hostfxr_initialize_for_dotnet_command_line_fn`? Or are you loading multiple assemblies other than that one?

If you are not explicitly loading different assemblies (that is, you need to call into the one used for `hostfxr_initialize_for_dotnet_command_line_fn`), you could keep everything in the default ALC by using `get_function_pointer`. If you are loading other assemblies, we recommend shared dependencies live in the default ALC and components/plugins prevent those shared dependencies from being added to their .deps.json by setting [`Private=false` / `ExcludeAssets=runtime`](https://learn.microsoft.com/dotnet/core/tutorials/creating-app-with-plugin-support#reference-a-plugin-interface-from-a-nuget-package) on the reference.

**Comment by MansenC** [comment]
I've checked which context an assembly was loaded into using `Assembly.HostContext` which displayed 0 for both so I assumed they're in the same context. That they're in fact not is good to know. Seems like I found the fix by just trying around and using `hostfxr_run_app` combined with `get_function_pointer` after that. Luckily in that case I am only using that one assembly that I call into.

Thank you for the response and explanation, since everything is resolved I am closing this issue here
